src/fugw/solvers/utils.py

```diff
@@ -450,6 +449,539 @@ def solver_sinkhorn_sparse(
+                pi = get_K(alpha + eps * u.log(), beta + eps * v.log())
+                err = torch.norm(pi.sum(0) - wt)
+                if err < tol:
+                    if verbose:
+                        progress.console.log(
+                            f"Reached tolerance threshold: {err}"
+                        )
+                    break
+
+            if verbose:
+                progress.update(task, advance=1)
+
+            idx += 1
+
+    # Compute final transport plan
+    pi = get_pi(alpha, beta, u, v)
+
+    # Final update to potentials
+    alpha = alpha + eps * u.log()
+    beta = beta + eps * v.log()
+
+    return (alpha, beta), pi
+
+
+def solver_sinkhorn_stabilized_sparse(
+    cost,
+    ws,
+    wt,
+    eps,
+    init_duals=None,
+    numItermax=1000,
+    tol=1e-9,
+    eval_freq=20,
+    stabilization_threshold=1e3,
+    verbose=True,
+):
+    """
+    Stabilized sparse Sinkhorn algorithm following the structure of the dense
+    stabilized implementation but using sparse matrix operations.
+
+    Parameters
+    ----------
+    cost: torch.sparse_csr_tensor
+        Sparse cost matrix.
+    ws: torch.Tensor
+        Source weights.
+    wt: torch.Tensor
+        Target weights.
+    eps: float
+        Entropy regularization parameter.
+    init_duals: tuple of torch.Tensor, optional, defaults to None
+        Initial dual potentials.
+    numItermax: int, optional, defaults to 1000
+        Maximum number of iterations.
+    tol: float, optional, defaults to 1e-9
+        Tolerance threshold.
+    eval_freq: int, optional, defaults to 20
+        Frequency at which to evaluate the tolerance threshold.
+    stabilization_threshold: float, optional, defaults to 1e3
+        Threshold for numerical stabilization.
+    verbose: bool, optional, defaults to False
+        Whether to display progress.
+
+    Returns
+    -------
+    (alpha, beta): tuple of torch.Tensor
+        Dual potentials.
+    pi: torch.sparse_csr_tensor
+        Optimal transport plan.
+    """
+
+    # Initialize scaling vectors to ones
+    u = torch.ones_like(ws) / len(ws)
+    v = torch.ones_like(wt) / len(wt)
+
+    # Initialize the dual potentials
+    if init_duals is None:
+        alpha = torch.zeros_like(ws)
+        beta = torch.zeros_like(wt)
+    else:
+        alpha, beta = init_duals
+
+    # Set up sparse matrix operations
+    crow_indices = cost.crow_indices()
+
+    # Prepare sparse matrices for efficient computation
+    cost_csc = cost.to_sparse_csc()
+    ccol_indices = cost_csc.ccol_indices()
+    T = torch.sparse_csc_tensor(
+        cost_csc.ccol_indices(),
+        cost_csc.row_indices(),
+        # Add 1 to arange so that first coefficient is not 0
+        torch.arange(cost_csc.values().shape[0]) + 1,
+        size=cost_csc.size(),
+        device=cost_csc.device,
+    ).to_sparse_csr()
+    csc_to_csr = T.values() - 1
+
+    # Compute kernel matrix K in sparse format
+    def get_K_sparse(alpha, beta):
```

Comment:
Don't you want to put these functions outside of the other one (for readability) ?